File: example/lib/screens/biologin.dart
import 'package:flutter/material.dart';
import 'package:flutter_amanisdk/amani_sdk.dart';
import 'package:flutter_amanisdk/common/models/android/auto_selfie_settings.dart';
import 'package:flutter_amanisdk/common/models/android/pose_estimation_settings.dart';
import 'package:flutter_amanisdk/common/models/ios/auto_selfie_settings.dart';
import 'package:flutter_amanisdk/common/models/ios/pose_estimation_settings.dart';

class BioLogin extends StatefulWidget {
  const BioLogin({Key? key}) : super(key: key);

  @override
  State<BioLogin> createState() => _BioLoginState();
}

class _BioLoginState extends State<BioLogin> {
  final _bioLogin = AmaniSDK().getBioLogin();
  String _customerID = "";

  Future<void> initBioLogin() async {
    await _bioLogin.init(
        server: "server",
        token: "bio login token from the biologin server",
        customerId: _customerID,
        attemptId: "string of attempt id which you should generate");
  }

  @override
  void initState() {
    super.initState();
    initBioLogin();
  }

  final AndroidPoseEstimationSettings _androidPoseEstimationSettings =
      AndroidPoseEstimationSettings(
    poseCount: 3,
    animationDuration: 500,
    faceNotInside: "Your face is not inside the area",
    faceNotStraight: "Your face is not straight",
    faceIsTooFar: "Your face is too far",
    holdPhoneVertically: "Please hold stable",
    alertTitle: "Verification Failed",
    alertDescription: "Failed 1",
    alertTryAgain: "Try again",
    turnRight: "→",
    turnLeft: "←",
    turnUp: "↑",
    turnDown: "↓",
    lookStraight: "Look straight",
    mainGuideVisibility: true,
    secondaryGuideVisibility: true,
  );

  final IOSPoseEstimationSettings _iosPoseEstimationSettings =
      IOSPoseEstimationSettings(
    faceIsOk: "Please hold stable",
    notInArea: "Please align your face to the area",
    faceTooSmall: "Your face is in too far",
    faceTooBig: "Your face is in too close",
    completed: "Verification Completed",
    turnRight: "→",
    turnLeft: "←",
    turnUp: "↑",
    turnDown: "↓",
    lookStraight: "Look straight",
    errorMessage:
        "Please complete the steps while your face is aligned to the area",
    tryAgain: "Try again",
    errorTitle: "Verification Failure",
    confirm: "Confirm",
    next: "Next",
    holdPhoneVertically: "Please hold the phone straight",
    informationScreenDesc1:
        "To start verification, align your face with the area",
    informationScreenDesc2: "",
    informationScreenTitle: "Selfie Verification Instructions",
    wrongPose: "Your face must be straight",
    descriptionHeader:
        "Please make sure you are doing the correct pose and your face is aligned with the area",
    appBackgroundColor: "000000",
    appFontColor: "ffffff",
    primaryButtonBackgroundColor: "ffffff",
    primaryButtonTextColor: "000000",
    ovalBorderColor: "ffffff",
    ovalBorderSuccessColor: "00ff00",
    poseCount: "3",
    mainGuideVisibility: "true",
    secondaryGuideVisibility: "true",
    buttonRadious: "10",
    manualCropTimeout: 30,
  );

  final IOSAutoSelfieSettings _iOSAutoSelfieSettings = IOSAutoSelfieSettings(
      faceIsOk: "Please hold stable",
      notInArea: "Please align your face with the area",
      faceTooSmall: "Your face is in too far",
      faceTooBig: "Your face is in too close",
      completed: "All OK!",
      appBackgroundColor: "000000",
      appFontColor: "ffffff",
      primaryButtonBackgroundColor: "ffffff",
      ovalBorderSuccessColor: "00ff00",
      ovalBorderColor: "ffffff",
      countTimer: "3",
      manualCropTimeout: 30);

  final AndroidAutoSelfieSettings _androidAutoSelfieSettings =
      AndroidAutoSelfieSettings(
          textSize: 16,
          counterVisible: true,
          counterTextSize: 21,
          manualCropTimeout: 30,
          distanceText: "Please align your face with the area",
          faceNotFoundText: "No faces found",
          restartText: "Process failed, restarting...",
          stableText: "Please hold stable");

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar:
          AppBar(backgroundColor: Colors.red, title: const Text("BioLogin")),
      body: Center(
        child: Column(
          children: [
            TextField(
              decoration: const InputDecoration(
                  labelText: "Customer ID", border: UnderlineInputBorder()),
              onChanged: (customerID) {
                setState(() {
                  _customerID = customerID;
                });
              },
            ),
            OutlinedButton(
                onPressed: () {
                  _bioLogin.startWithAutoSelfie(
                      iosAutoSelfieSettings: _iOSAutoSelfieSettings,
                      androidAutoSelfieSettings: _androidAutoSelfieSettings);
                },
                child: const Text("Start AutoSelfie BioLogin")),
            OutlinedButton(
                onPressed: () {
                  _bioLogin
                      .iOSStartWithPoseEstimation(
                          settings: _iosPoseEstimationSettings)
                      .then((imageData) {
                    // do something with imageData
                  });
                },
                child: const Text("Start PoseEstimation BioLogin")),
            OutlinedButton(
                onPressed: () {
                  _bioLogin.androidStartWithPoseEstimation(
                      settings: _androidPoseEstimationSettings,
                      onSuccessCallback: (imageData) {
                        // render the image with Image.memory(imageData)
                      },
                      onFailureCallback: (failureParams) {
                        print(failureParams["reason"] +
                            failureParams["currentAttempt"]);
                      });
                },
                child: const Text("Start PoseEstimation BioLogin (Android)")),
            OutlinedButton(
                onPressed: () {
                  _bioLogin.startWithManualSelfie(
                      androidSelfieDescriptionText: "Hi");
                },
                child: const Text("Start Manual Selfie BioLogin")),
          ],
        ),
      ),
    );
  }
}

File: ios/Classes/modules/BioLogin.swift
//
//  BioLogin.swift
//  flutter_amanisdk_v2
//
//  Created by Deniz Can on 22.11.2022.
//
import AmaniSDK
import UIKit
import Flutter

class BioLogin {
    private let vc = UIApplication.shared.windows.last?.rootViewController!
    private let module = Amani.sharedInstance.bioLogin()
    private var sdkView: SDKView!
    private var customerId: String!
    private var attemptId: String!
    private var source = 3
    private var comparison_adapter = 2
    private var server: String!
    private var token: String!
    
    // Locking the initializer
    private init() {}
    
    public static let shared = BioLogin()
    
    public func initBioLogin(server: String, token: String, customerId: String, attemptId: String, source: Int?, comparisonAdapter: Int?, result: @escaping FlutterResult) {
        self.customerId = String(customerId)
        self.source = source ?? self.source
        self.comparison_adapter = comparisonAdapter ?? self.comparison_adapter
        self.server = server
        self.token = token
        self.attemptId = attemptId
        
        module.setParams(server: server, token: token, customer_id: customerId, source: source ?? 3, comparison_adapter: comparisonAdapter ?? 2, attempt_id: attemptId)
        result(nil)
    }
    
    
    public func startWithAutoSelfie(settings: AutoSelfieSettings, result: @escaping FlutterResult) {
        module.setParams(server: server, token: token, customer_id: customerId, source: source, comparison_adapter: comparison_adapter, attempt_id: attemptId, biologintype: .autoSelfie)
        module.setSelfieMessages(key: .faceIsOk, value: settings.faceIsOk)
        module.setSelfieMessages(key: .notInArea, value: settings.notInArea)
        module.setSelfieMessages(key: .faceTooBig, value: settings.faceTooBig)
        module.setSelfieMessages(key: .faceTooSmall, value: settings.faceTooSmall)
        module.setSelfieMessages(key: .completed, value: settings.completed)
        
        module.setSelfieColor(key: .appBackgroundColor, value: settings.appBackgroundColor)
        module.setSelfieColor(key: .appFontColor, value: settings.appFontColor)
        module.setSelfieColor(key: .primaryButtonBackgroundColor, value: settings.primaryButtonBackgroundColor)
        module.setSelfieColor(key: .ovalBorderSuccessColor, value: settings.ovalBorderSuccessColor)
        module.setSelfieColor(key: .ovalBorderColor, value: settings.ovalBorderColor)
        module.setSelfieColor(key: .countTimer, value: settings.countTimer)
        
        startModule(result: result)
    }
    
    public func startWithPoseEstimation(settings: PoseEstimationSettings, result: @escaping FlutterResult) {
        module.setParams(server: server, token: token, customer_id: customerId, source: source, comparison_adapter: comparison_adapter, attempt_id: attemptId, biologintype: .poseEstimation)
        module.setPoseEstimationColors(screenConfig: [
            .appBackgroundColor: settings.appBackgroundColor,
            .appFontColor: settings.appFontColor,
            .primaryButtonBackgroundColor: settings.primaryButtonBackgroundColor,
            .primaryButtonTextColor: settings.primaryButtonTextColor,
            .ovalBorderColor: settings.ovalBorderColor,
            .ovalBorderSuccessColor: settings.ovalBorderSuccessColor,
            .poseCount: settings.poseCount,
            .mainGuideVisibility: settings.mainGuideVisibility,
            .secondaryGuideVisibility: settings.secondaryGuideVisibility,
            .buttonRadius: settings.buttonRadious,
        ])
        module.setPoseEstimationMessages(infoMessages: [
            .faceIsOk: settings.faceIsOk,
            .notInArea: settings.notInArea,
            .faceTooSmall: settings.faceTooSmall,
            .faceTooBig: settings.faceTooBig,
            .completed: settings.completed,
            .turnRight: settings.turnRight,
            .turnLeft: settings.turnLeft,
            .turnUp: settings.turnUp,
            .turnDown: settings.turnDown,
            .lookStraight: settings.lookStraight,
            .errorMessage: settings.errorMessage,
            .errorTitle: settings.errorTitle,
            .confirm: settings.confirm,
            .next: settings.next,
            .holdPhoneVertically: settings.holdPhoneVertically,
            .informationScreenDesc1: settings.informationScreenDesc1,
            .informationScreenDesc2: settings.informationScreenDesc2,
            .informationScreenTitle: settings.informationScreenTitle,
            .wrongPose: settings.wrongPose,
            .descriptionHeader: settings.descriptionHeader
        ])
  
        startModule(result: result)
    }
    
    public func startWithManualSelfie(result: @escaping FlutterResult) {
        module.setParams(server: server, token: token, customer_id: customerId, source: source, comparison_adapter: comparison_adapter, attempt_id: attemptId, biologintype: .manualSelfie)
        startModule(result: result)
    }
    
    public func upload(result: @escaping FlutterResult) {
      module.upload { (isSuccess, error) in
        if let success = isSuccess, success {
          result(success)
        } else {
          if let error = error {
            let errorsDict = error.map {
              $0.toDictionary()
            }
            result(FlutterError(code: "30010", message: "Upload result errors", details: errorsDict))
          } else {
            result(FlutterError(code: "30011", message: "Upload result returning with nil value", details: nil))
          }
        }
      }
    }
    
    private func startModule(result: @escaping FlutterResult) {
        do {
            let moduleView = try module.start { image in
                let data = image.pngData()!
                result(FlutterStandardTypedData(bytes: data))
                DispatchQueue.main.async {
                    self.sdkView.removeFromSuperview()
                }
            }
          sdkView = SDKView(sdkView: moduleView!)
          sdkView.start(on: vc!)
          sdkView.setupBackButton(on: moduleView!)
        } catch let err {
            result(FlutterError(code: "30007", message: err.localizedDescription, details: nil))
        }
    }
    
}
